# 1141_社會資料分析 作業七
學生：劉晏成

以「2018台北大學社會系三鶯調查」（brilliant_working.sav）檔案回答以下問題：

1. 以收入（inc）作為依變項，性別（sex）為自變項，檢驗教育年數（eduy）是否可以當做上述二者的中介變項。  
如果有中介關係，計算 (bc - b‘c) / bc ，顯示中介關係占原關係的百分比
2. 以收入（inc）作為依變項，教育年數（eduy）做為自變項、人際網路（net）作為調節變項，檢驗教育年數對收入的影響是否受到人際網路調節。

## Notes
1. 請先做整體模型的描述(R平方、顯著性等)。
2. 第一題請列出迴歸方程式與計算過程。若無需繼續檢視中介效果請以文字說明(請參考本週簡報p.27)。
3. 第二題交互作用項顯著與否的後續作法不同，請操考調節變項的示範一與練習一，但不需附上折線圖。
4. 請看清楚自變項、依變項、中介、調節分別是哪個變項，也請看清楚P值以免做出錯誤判斷。

In [1]:
from load import load_sav
import pandas as pd

df = load_sav("../data/brilliant_working.sav")

In [2]:
variable_value_labels = df.attrs['variable_value_labels']
column_names = df.attrs['column_names']
column_names_to_labels = df.attrs['column_names_to_labels']

## Q1: 以收入（inc）作為依變項，性別（sex）為自變項，檢驗教育年數（eduy）是否可以當做上述二者的中介變項。
如果有中介關係，計算 (bc - b‘c) / bc ，顯示中介關係占原關係的百分比

In [3]:
df_1 = df[['inc', 'sex', 'eduy']].copy()

檢查遺漏值狀況

In [4]:
from statistic.missing_value import check_missing_value

check_missing_value(df_1)

# of Missing Values
inc                     4
sex                     0
eduy                    0

發現有遺漏值，處理遺漏值

In [5]:
df_1_cleared = df_1.dropna()
check_missing_value(df_1_cleared)

# of Missing Values
inc                     0
sex                     0
eduy                    0

In [6]:
from statistic.linear_regression import mediation_analysis
reports = mediation_analysis(X = df_1_cleared['sex'], Y = df_1_cleared['inc'], Z = df_1_cleared['eduy'])

檢視 $Y = b_cX + a_1 $ 迴歸結果

In [7]:
reports['total_effect_summary']

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    inc   R-squared:                       0.048
Model:                            OLS   Adj. R-squared:                  0.047
Method:                 Least Squares   F-statistic:                     43.67
Date:                Thu, 11 Dec 2025   Prob (F-statistic):           6.81e-11
Time:                        00:09:55   Log-Likelihood:                -10119.
No. Observations:                 864   AIC:                         2.024e+04
Df Residuals:                     862   BIC:                         2.025e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       2.992e+04   1396.482     21.427      0.000    2.72e+04    3.27e+04
sex          1.33e+04   2012.545      6.608      0.000    9349.217    1.72e+04
==============================================================================
Omnibus:                      217.946   Durbin-Watson:                   1.964
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              499.360
Skew:                           1.357   Prob(JB):                    3.68e-109
Kurtosis:                       5.551   Cond. No.                         2.58
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

小結：$b_c$ 為13300， P value 為0；由於P value < 0.05 ，具統計顯著性

檢視 $Z = b_aX + a_2 $ 迴歸結果

In [8]:
reports['mediator_summary']

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   eduy   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                 -0.000
Method:                 Least Squares   F-statistic:                    0.8234
Date:                Thu, 11 Dec 2025   Prob (F-statistic):              0.364
Time:                        00:09:55   Log-Likelihood:                -2367.4
No. Observations:                 864   AIC:                             4739.
Df Residuals:                     862   BIC:                             4748.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         13.4509      0.177     75.886      0.000      13.103      13.799
sex            0.2318      0.255      0.907      0.364      -0.270       0.733
==============================================================================
Omnibus:                       30.342   Durbin-Watson:                   1.773
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               32.551
Skew:                          -0.464   Prob(JB):                     8.54e-08
Kurtosis:                       2.791   Cond. No.                         2.58
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

小結：$b_a$ 為 0.2318， P value 為0.364；由於P value > 0.05 ，不具統計顯著性

### Answer
整體模型的R Square 為0.048，具統計顯著性。

由於 $b_a$ 不具統計顯著性，中介檢驗不通過，停止檢驗。教育年份對收入與性別無中介作用。

## Q2: 以收入（inc）作為依變項，教育年數（eduy）做為自變項、人際網路（net）作為調節變項，檢驗教育年數對收入的影響是否受到人際網路調節。

In [9]:
df_2 = df[['inc', 'eduy', 'net']].copy()

檢查遺漏值狀況

In [10]:
from statistic.missing_value import check_missing_value

check_missing_value(df_2)

# of Missing Values
inc                     4
eduy                    0
net                     0

發現有遺漏值，處理遺漏值

In [11]:
df_2_cleared = df_2.dropna()
check_missing_value(df_2_cleared)

# of Missing Values
inc                     0
eduy                    0
net                     0

In [12]:
from statistic.linear_regression import moderator_analysis
reports = moderator_analysis(X = df_2_cleared['eduy'], Y = df_2_cleared['inc'], Z = df_2_cleared['net'])

首先檢驗交互作用，檢視 $Y = b_1X + b_2Z + b_3XZ + a $ 迴歸結果

In [13]:
reports['interaction_summary']

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    inc   R-squared:                       0.079
Model:                            OLS   Adj. R-squared:                  0.076
Method:                 Least Squares   F-statistic:                     24.50
Date:                Thu, 11 Dec 2025   Prob (F-statistic):           3.21e-15
Time:                        00:12:42   Log-Likelihood:                -10105.
No. Observations:                 864   AIC:                         2.022e+04
Df Residuals:                     860   BIC:                         2.024e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       8001.2375   8322.140      0.961      0.337   -8332.846    2.43e+04
eduy        1579.0175    588.183      2.685      0.007     424.576    2733.459
net         -250.4052   1424.626     -0.176      0.861   -3046.556    2545.746
eduy*net     123.4837    104.640      1.180      0.238     -81.896     328.863
==============================================================================
Omnibus:                      229.740   Durbin-Watson:                   1.993
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              553.264
Skew:                           1.406   Prob(JB):                    7.25e-121
Kurtosis:                       5.732   Cond. No.                         652.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

交互作用項P Value 為0.238 ，由於P Value > 0.05 ，故交互作用效果未達統計顯著

接下來去除交互作用項，檢視檢視 $Y = b_1X + b_2Z + a $ 迴歸結果

In [14]:
reports['main_effects_summary']

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    inc   R-squared:                       0.077
Model:                            OLS   Adj. R-squared:                  0.075
Method:                 Least Squares   F-statistic:                     36.04
Date:                Thu, 11 Dec 2025   Prob (F-statistic):           9.28e-16
Time:                        00:12:42   Log-Likelihood:                -10106.
No. Observations:                 864   AIC:                         2.022e+04
Df Residuals:                     861   BIC:                         2.023e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       -247.0795   4518.036     -0.055      0.956   -9114.732    8620.573
eduy        2197.2922    267.384      8.218      0.000    1672.492    2722.093
net         1361.5560    404.661      3.365      0.001     567.318    2155.794
==============================================================================
Omnibus:                      228.265   Durbin-Watson:                   1.985
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              545.803
Skew:                           1.400   Prob(JB):                    3.02e-119
Kurtosis:                       5.706   Cond. No.                         68.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Answer
教育年數與人際網路對收入無顯著交互作用影響；故就整體模型呈現教育年數、人際網路與收入的關係。

模型R Square 為0.077，模型P Value < 0.05 ，模型具統計顯著性

分別檢視教育年數以及人際網路，兩者P Value 皆 < 0.05 ，兩者對於收入影響皆具備統計顯著性。